File: src/main/java/me/itzisonn_/meazy/parser/ast/statement/FunctionDeclarationStatement.java
package me.itzisonn_.meazy.parser.ast.statement;

import lombok.Getter;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.registry.Registries;
import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.modifier.Modifier;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.ParameterExpression;
import me.itzisonn_.meazy.parser.modifier.Modifiers;
import me.itzisonn_.meazy.runtime.environment.*;
import me.itzisonn_.meazy.runtime.variable_value.VariableValue;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.lang.constant.ClassDesc;
import java.lang.constant.ConstantDescs;
import java.lang.constant.MethodTypeDesc;
import java.lang.reflect.AccessFlag;
import java.util.*;

@Getter
@NullMarked
public class FunctionDeclarationStatement extends ModifierStatement implements DeclarationStatement {
    private final String id;
    @Nullable
    private final String classId;
    private final List<ParameterExpression> parameters;
    private final List<LocalStatement> body;
    @Nullable
    private final DataType returnDataType;
    @Nullable
    private final Expression returnDataTypeValue;
    @Nullable
    private FunctionEnvironment functionEnvironment;

    public FunctionDeclarationStatement(
            Set<Modifier> modifiers, String id, @Nullable String classId, List<ParameterExpression> parameters,
            List<LocalStatement> body, @Nullable DataType returnDataType, @Nullable Expression returnDataTypeValue
    ) {
        super(modifiers);
        this.id = id;
        this.classId = classId;
        this.parameters = parameters;
        this.body = body;
        this.returnDataType = returnDataType;
        this.returnDataTypeValue = returnDataTypeValue;
    }

    public FunctionDeclarationStatement(Set<Modifier> modifiers, String id, List<ParameterExpression> parameters, List<LocalStatement> body, @Nullable DataType returnDataType) {
        this(modifiers, id, null, parameters, body, returnDataType, null);
    }

    @Override
    public void declare(Environment environment) { //TODO check whether abstract function isn't overridden
        if (!(environment instanceof FunctionDeclarationEnvironment functionDeclarationEnvironment)) {
            throw new RuntimeException("CANT DECLARE FUNCTION HERE TODO");
        }

        boolean isShared = modifiers.contains(Modifiers.SHARED()) || environment instanceof FileEnvironment;

        FunctionEnvironment functionEnvironment = Registries.FUNCTION_ENVIRONMENT_FACTORY.getEntry().getValue().create(
                functionDeclarationEnvironment, null, null, id, parameters,
                returnDataType, isShared, modifiers
        );

        functionDeclarationEnvironment.declareFunction(functionEnvironment);
        this.functionEnvironment = functionEnvironment;

        if (modifiers.contains(Modifiers.ABSTRACT())) return;

        for (LocalStatement localStatement : body) {
            if (localStatement.alwaysReturns()) return;
        }

        if (returnDataType == null) {
            body.add(new ReturnStatement(null));
            return;
        }

        throw new RuntimeException("Function with id " + id + " doesn't always return a value");
    }

    @Override
    public void resolve(Environment environment) {
        if (functionEnvironment == null) {
            throw new RuntimeException("Function isn't declared TODO");
        }

        DataType returnDataType;
        if (functionEnvironment.getReturnDataType() != null) returnDataType = functionEnvironment.getReturnDataType();
        else if (returnDataTypeValue != null) {
            returnDataType = returnDataTypeValue.getType(environment, this);
            functionEnvironment.setReturnDataType(returnDataType);
        }
        else returnDataType = null;

        if (returnDataType != null) returnDataType.resolve(environment);

        functionEnvironment.getParameters().forEach(parameter -> parameter.getDataType().resolve(environment));
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, ProgramUnit parent) {
        if (functionEnvironment == null) {
            throw new RuntimeException("Declared function is unresolved TODO");
        }

        UUID startLabel = instructionsSet.createLabel();
        UUID endLabel = instructionsSet.createLabel();
        functionEnvironment.setStartLabel(startLabel);
        functionEnvironment.setEndLabel(endLabel);

        if (functionEnvironment.getParent() instanceof ClassEnvironment classEnvironment) {
            if (classEnvironment.getBaseClass() != null) {
                EnvironmentUtils.getClassEnvironment(classEnvironment, classEnvironment.getBaseClass()).orElseThrow()
                        .getFunctionRecursively(this.functionEnvironment.getId(), this.functionEnvironment.getParameters().stream().map(ParameterExpression::getDataType).toList())
                        .ifPresent(f -> {
                            if (!f.getModifiers().contains(Modifiers.OPEN()) && !f.getModifiers().contains(Modifiers.ABSTRACT())) {
                                throw new RuntimeException("Can't override non-open function " + id);
                            }
                            if (!this.functionEnvironment.getModifiers().contains(Modifiers.OVERRIDE())) throw new RuntimeException("Must specify override keyword on function " + id);
                        });
            }

            for (ClassDesc interfaceClassDesc : classEnvironment.getInterfaces()) {
                EnvironmentUtils.getClassEnvironment(classEnvironment, interfaceClassDesc).orElseThrow()
                        .getFunctionRecursively(this.functionEnvironment.getId(), this.functionEnvironment.getParameters().stream().map(ParameterExpression::getDataType).toList())
                        .ifPresent(f -> {
                            if (!f.getModifiers().contains(Modifiers.OPEN()) && !f.getModifiers().contains(Modifiers.ABSTRACT())) {
                                throw new RuntimeException("Can't override non-open function " + id);
                            }
                            if (!this.functionEnvironment.getModifiers().contains(Modifiers.OVERRIDE())) throw new RuntimeException("Must specify override keyword on function " + id);
                        });
            }
        }

        boolean isShared = functionEnvironment.isShared();
        DataType returnDataType = this.functionEnvironment.getReturnDataType();

        MethodTypeDesc methodTypeDesc = MethodTypeDesc.of(
                returnDataType == null ? ConstantDescs.CD_void : returnDataType.getClassDesc(),
                this.functionEnvironment.getParameters().stream().map(p -> p.getDataType().getClassDesc()).toList()
        );

        Set<AccessFlag> accessFlags = new HashSet<>();
        if (this.functionEnvironment.getModifiers().contains(Modifiers.PRIVATE())) accessFlags.add(AccessFlag.PRIVATE);
        else if (this.functionEnvironment.getModifiers().contains(Modifiers.PROTECTED())) accessFlags.add(AccessFlag.PROTECTED);
        else accessFlags.add(AccessFlag.PUBLIC);

        if (isShared) accessFlags.add(AccessFlag.STATIC);
        if (this.functionEnvironment.getModifiers().contains(Modifiers.ABSTRACT())) accessFlags.add(AccessFlag.ABSTRACT);
        else if (!this.functionEnvironment.getModifiers().contains(Modifiers.OPEN()) && !(functionEnvironment.getParent() instanceof ClassEnvironment classEnvironment && classEnvironment.isInterface())) {
            accessFlags.add(AccessFlag.FINAL);
        }

        instructionsSet.withMethod(
                this.functionEnvironment.getId(),
                methodTypeDesc,
                accessFlags,
                bodyInstructions -> {
                    bodyInstructions.initLabel(startLabel);
                    bodyInstructions.initLabel(endLabel);

                    for (ParameterExpression parameter : this.functionEnvironment.getParameters()) {
                        VariableValue parameterValue = functionEnvironment.declareVariable(
                                parameter.getId(),
                                parameter.getDataType(),
                                parameter.isConstant(),
                                null
                        );

                        bodyInstructions.setLocalName(parameterValue.getSlot(), parameter.getId(), parameter.getDataType().getClassDesc(), startLabel, endLabel);
                    }

                    bodyInstructions.bindLabel(startLabel);
                    for (Statement statement : body) {
                        statement.emit(bodyInstructions, functionEnvironment, this);
                    }
                    bodyInstructions.bindLabel(endLabel);
                }
        );
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/statement/FunctionDeclarationStatement.kt
package me.itzisonn_.meazy.parser.ast.statement

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.ast.ParentMap
import me.itzisonn_.meazy.runtime.data.DataType
import me.itzisonn_.meazy.runtime.data.Parameter
import me.itzisonn_.meazy.runtime.data.modifier.Modifier
import me.itzisonn_.meazy.runtime.data.modifier.Modifiers
import me.itzisonn_.meazy.runtime.data.symbol.FunctionSymbol
import me.itzisonn_.meazy.runtime.environment.*
import me.itzisonn_.meazy.runtime.environment.declaration.FunctionDeclarationEnvironment
import java.lang.constant.ConstantDescs
import java.lang.constant.MethodTypeDesc
import java.lang.reflect.AccessFlag

class FunctionDeclarationStatement(
    modifiers: Set<Modifier>,
    val id: String,
    val classId: String?,
    val parameters: List<Parameter>,
    body: List<LocalStatement>,
    val returnDataType: DataType?
) : ModifierStatement(modifiers), DeclarationStatement {
    val body = body.toMutableList()
    private lateinit var symbol: FunctionSymbol

    override val children = body.toSet()

    constructor(
        modifiers: Set<Modifier>, id: String, parameters: List<Parameter>,
        body: List<LocalStatement>, returnDataType: DataType?
    ) : this(
        modifiers, id, null, parameters, body, returnDataType
    )

    context(parents: ParentMap)
    override fun declare(environment: Environment) {
        //TODO check whether abstract function isn't overridden
        if (environment !is FunctionDeclarationEnvironment) {
            throw RuntimeException("CANT DECLARE FUNCTION HERE TODO")
        }

        val isShared = Modifiers.shared in modifiers || environment is FileEnvironment

        val functionEnvironment = FunctionEnvironment(
            environment, null, null, id, parameters,
            returnDataType, isShared, modifiers
        )

        symbol = FunctionSymbol(
            id, parameters, returnDataType, modifiers, functionEnvironment
        )
        environment.declareFunction(symbol)

        if (Modifiers.abstract in modifiers) return

        for (localStatement in body) {
            if (localStatement.alwaysReturns()) return
        }

        if (returnDataType == null) {
            body.add(ReturnStatement(null))
            return
        }

        throw RuntimeException("Function with id $id doesn't always return a value")
    }

    context(parents: ParentMap)
    override fun resolve(environment: Environment) {
        val functionEnvironment = symbol.environment
        functionEnvironment.returnDataType?.resolve(environment)
        functionEnvironment.parameters.forEach { it.dataType.resolve(environment) }
    }

    context(parents: ParentMap)
    override fun emit(instructions: InstructionsSet, environment: Environment) {
        val functionEnvironment = symbol.environment

        val startLabel = instructions.createLabel()
        val endLabel = instructions.createLabel()
        functionEnvironment.setStartLabel(startLabel)
        functionEnvironment.setEndLabel(endLabel)

        if (functionEnvironment.getParent() is ClassEnvironment) {
            val classEnvironment = functionEnvironment.getParent() as ClassEnvironment
            val baseClass = classEnvironment.baseClass

            if (baseClass != null) {
                val cls = classEnvironment.getClass(baseClass)
                    ?: error("Unknown base class $baseClass")

                cls.getFunctionRecursively(
                    functionEnvironment.id, functionEnvironment.parameters.map { it.dataType }
                )
                    ?.let { f ->
                        if (Modifiers.open !in f.modifiers && Modifiers.abstract !in f.modifiers) {
                            throw RuntimeException("Can't override non-open function $id")
                        }
                        if (Modifiers.override !in functionEnvironment.modifiers) throw RuntimeException(
                            "Must specify override keyword on function $id"
                        )
                    }
            }

            for (interfaceClassDesc in classEnvironment.interfaces) {
                val cls = classEnvironment.getClass(interfaceClassDesc)
                    ?: error("Unknown interface $interfaceClassDesc")

                cls.getFunctionRecursively(
                    functionEnvironment.id, functionEnvironment.parameters.map { it.dataType }
                )
                    ?.let { f ->
                        if (Modifiers.open !in f.modifiers && Modifiers.abstract !in f.modifiers) {
                            throw RuntimeException("Can't override non-open function $id")
                        }
                        if (Modifiers.override !in functionEnvironment.modifiers) throw RuntimeException(
                            "Must specify override keyword on function $id"
                        )
                    }
            }
        }

        val isShared = functionEnvironment.isShared
        val returnDataType = functionEnvironment.returnDataType

        val methodTypeDesc = MethodTypeDesc.of(
            returnDataType?.classDesc ?: ConstantDescs.CD_void,
            functionEnvironment.parameters.map { it.dataType.classDesc }
        )

        val accessFlags = mutableSetOf<AccessFlag>()
        if (Modifiers.private in functionEnvironment.modifiers) accessFlags.add(AccessFlag.PRIVATE)
        else if (Modifiers.protected in functionEnvironment.modifiers) accessFlags.add(AccessFlag.PROTECTED)
        else accessFlags.add(AccessFlag.PUBLIC)

        if (isShared) accessFlags.add(AccessFlag.STATIC)
        if (Modifiers.abstract in functionEnvironment.modifiers) accessFlags.add(AccessFlag.ABSTRACT)
        else if (Modifiers.open !in functionEnvironment.modifiers &&
            !(functionEnvironment.getParent() is ClassEnvironment && (functionEnvironment.getParent() as ClassEnvironment).isInterface)) {
            accessFlags.add(AccessFlag.FINAL)
        }

        instructions.withMethod(
            functionEnvironment.id,
            methodTypeDesc,
            accessFlags
        ) {
            initLabel(startLabel)
            initLabel(endLabel)

            for (parameter in functionEnvironment.parameters) {
                val parameterValue = functionEnvironment.declareVariable(
                    parameter.id,
                    parameter.dataType,
                    parameter.isConstant,
                    null,
                    setOf()
                )

                setLocalName(
                    parameterValue.slot,
                    parameter.id,
                    parameter.dataType.classDesc,
                    startLabel,
                    endLabel
                )
            }

            bindLabel(startLabel)
            for (statement in body) {
                statement.emit(this, functionEnvironment)
            }
            bindLabel(endLabel)
        }
    }
}
